Ros2 safety area manager

MarlonRiv opened the pull request:
## Control Manager

Refactored `ControlManager` to use `SafetyAreaManager` services instead of internal safety zone logic: - Reference validation services still present in control manager but calling the safety area manager services. This to be retro compatible with the current API. - Routines are being passed to trackers using the common handlers.

## Safety Area Manager

Dedicated manager that handles safety zone initialization and management.
Retro-compatible with previous world file, extra fields for obstacles (optional):


   <details>
  <summary>   New world file </summary>

```yaml

world_origin:

  units: "LATLON"

  origin_x: 47.397743
  origin_y: 8.545594

safety_area:

  enabled: true

  horizontal: 

    frame_name: "world_origin"

    points: [
      -53.113258, -52.640007,
      -45.805252, 47.673401,
      46.886742, 47.359993,
      46.886742, -52.640007,
    ]
  vertical:

    frame_name: "world_origin"

    max_z: 14.999985
    min_z: 0.0
    
obstacles:
    present: true
    count: 1

    obstacle_0:

      horizontal:

        frame_name: "world_origin"

        points: [
          -8.658495, 29.781685,
          -9.151301, 12.203030,
          -24.905661, 12.338413,
          -25.132828, 29.651335,
        ]

      vertical:

        frame_name: "world_origin"
        max_z: 7.400000
        min_z: -0.000007
```

 Refactored safety area having two main components: border and obstacles. 

 ### Border
    - Defined as a 3D prism using a horizontal polygon and vertical limits (min_z, max_z).
    - Same structure as previous safety zone but now in its own section given the new obstacle component.
 ### Obstacles;
    - Multiple obstacles can be defined within the safety area.
    - They correspond to the safety zone, but can be defined in different coordinate frames.

   </details>


## Safety Area Manager API
The `SafetyAreaManager` exposes services to interact with the safety area:

### Services

| Service | Description |
|---------|-------------|
| Set safety border | Allows dynamic reconfiguration of the safety border during runtime. Define a new border polygon with XY points and height (min_z, max_z). Options include `keep_obstacles` to retain existing obstacles and `update_world_origin` to update the world origin with the center of the new safety border. |
| Set obstacle | Sets an obstacle in the safety area with a custom-sized prism. |
| Add obstacle | Adds a new obstacle to the safety area. Requires only position; creates a default square obstacle with fixed size. |
| Toggle safety area | Enables or disables the safety area. |
| Validate 2D/3D point | Validates whether a 2D or 3D point is inside the safety area. |
| Validate path to point 2D/3D | Validates a path to a 2D or 3D point within the safety area. |
| Get max Z | Gets the maximum allowed Z coordinate at a given XY position. |
| Get min Z | Gets the minimum allowed Z coordinate at a given XY position. |
| Is safety area enabled | Checks whether the safety area is currently enabled. |

>[!NOTE] 
> When the world origin is updated using the `estimation_manager/set_world_origin` service, the safety border coordinates are recomputed based on the new origin. However, the `latlon_origin` (global coordinates) of the safety border will remain consistent, even if the previous area was initially defined using the "world_origin" in the `horizontal_frame` field.
> If a change in the deployment location is needed (changing world origin), a new safety area must be explicitly set.

### Diagnostics publishing

The safety area manager publishes a diagnostic message detailing the current status of the safety area. This message includes information such as the system world origin, whether the safety area is enabled, if the UAV is within the bounds of the safety zone, and the current border and obstacle definitions.

<details>
  <summary> Example of the diagnostic message published by the Safety Area Manager </summary>

```yaml
---
stamp:
  sec: 1762262673
  nanosec: 190044850
uav_name: uav1
world_origin:
  units: LATLON
  x: 47.397743
  y: 8.545594
safety_area_enabled: true
position_valid_2d: true
position_valid_3d: false
border:
  horizontal_frame: world_origin
  vertical_frame: world_origin
  points:
  - x: -53.113258
    y: -52.640007
  - x: -45.805252
    y: 47.673401
  - x: 46.886742
    y: 47.359993
  - x: 46.886742
    y: -52.640007
  max_z: 14.999985
  min_z: 0.499985
obstacles_present: true
obstacles:
- horizontal_frame: world_origin
  vertical_frame: world_origin
  points:
  - x: -8.658495
    y: 29.781685
  - x: -9.151301
    y: 12.20303
  - x: -24.905661
    y: 12.338413
  - x: -25.132828
    y: 29.651335
  max_z: 7.4
  min_z: -7.0e-06
---
```
</details>

----

This PR depends on changes in the following repos with open PR's:
- #### mrs_msgs:
    https://github.com/ctu-mrs/mrs_msgs/tree/ros2_safety_area_manager

    With the messages and services required for the safety area manager. 
    
  PR:  https://github.com/ctu-mrs/mrs_msgs/pull/15

- #### mrs_lib:
    https://github.com/ctu-mrs/mrs_lib/tree/ros2_safety_area_manager

    With utility functions for safety area management like safety zone implementation, definition of new prism structure, and visualization publishers. 
    
  PR: https://github.com/ctu-mrs/mrs_lib/pull/68


- #### mrs_uav_autostart:
    Not dependent, but simplifies the starting position validation:

  PR: https://github.com/ctu-mrs/mrs_uav_autostart/pull/3

MarlonRiv committed:
Removed parts for safety area (WIP)

WIP: Need to implement the safety area manager, and find proper way to
share the safety zone with control manager so it can use it in the
Reference validations.

MarlonRiv committed:
(WIP) Working safety area manager

Minimal version without services, publishers, and other features.

It builds, but needs cleaning.

MarlonRiv committed:
Services implementation and cleaning.

Missing minor details

MarlonRiv committed:
Added safety area manager services into control manager and minor changes

MarlonRiv committed:
cleaning safety area manager

MarlonRiv committed:
Simplified the obstacles loading from file

MarlonRiv committed:
Merge branch 'ros2' into ros2_safety_area_manager

MarlonRiv committed:
renamed launch file given ros2 convention

MarlonRiv committed:
Polishing, improving according to changes in Prism and some other updgrades for safe access.

MarlonRiv committed:
Improving the border creation and move semantics

MarlonRiv committed:
formatting

MarlonRiv committed:
cleaning and using current horizontal frame for transformations

MarlonRiv committed:
Merge branch 'ros2' into ros2_safety_area_manager

MarlonRiv committed:
Refactor safety area manager to use mrs_lib::Node and fix control manager from merge

MarlonRiv committed:
Merge branch 'ros2' into ros2_safety_area_manager

MarlonRiv committed:
Add the update of world origin service into safety area manager

Service intended to be used when calling
EstimationManager/set_world_origin, to update for the safety area
manager and display up to date information in diagnostics and
transformation of the given area

MarlonRiv committed:
Recomputing safety area border and obstacles with a change of world origin

In scenarios where a safety border and obstacles are set to respect to
the world origin, if a change of world origin is done the safety area
stays in the global coordinate frame (latlon) position, and doesn't
follow the new world origin. Given that the safety area was set to
respect to the previous world origin.

MarlonRiv committed:
Improve the global zone creation and new feature to update world origin with the centroid of the safety border.

MarlonRiv committed:
Improve error handling

MarlonRiv committed:
Improving thread safety and more error handling

MarlonRiv committed:
Fix grammar and typos

MarlonRiv committed:
Renamed toggle safety area remapping and obstacle service name

MarlonRiv committed:
Add remappings for the rviz marker publishers

spurnvoj commented:
@MarlonRiv, didn't we discuss that the RViz publisher will be part of the RViz plugins and not in this repository?

MarlonRiv commented:
> @MarlonRiv, didn't we discuss that the RViz publisher will be part of the RViz plugins and not in this repository?

@spurnvoj  We are initializing the publisher in the safety area manager but the publisher itself it's defined in `mrs_lib::StaticEdgesVisualization`, not in this repo.

https://github.com/ctu-mrs/mrs_uav_managers/blob/ros2_safety_area_manager/src/safety_area_manager.cpp#L1288C3-L1291C139

Implementation in `mrs_lib`:
https://github.com/ctu-mrs/mrs_lib/blob/ros2_safety_area_manager/src/safety_zone/static_edges_visualization.cpp

---

If i am not mistaken we discussed about the RViz plugins repo for the additional feature of the dynamic/interactive markers -- as this is could be an add-on/util for the RViz interaction -- but not about the standard markers.

RViz plugins is not part of `uav_core`, so in case we would have all the publisher implementations there we would need to add it:
https://github.com/ctu-mrs/mrs_uav_core/tree/ros2/ros_packages

MarlonRiv committed:
Cleaning

Consistency fixes

MarlonRiv committed:
Merge branch 'ros2' into ros2_safety_area_manager

MarlonRiv committed:
Solved conflicts with latest formatting commits, and applied new formatting to changes

MarlonRiv committed:
Added shutdown callback to stop timer

MarlonRiv committed:
Applied new formatting

MarlonRiv committed:
Improved world config structure to be retrocompatible

MarlonRiv committed:
Update with latest ros2 branch

MarlonRiv committed:
Building of safety area manager interface library

klaxalk commented:
@MarlonRiv hey, I am getting this in Rviz when visualizing the safety area coordinate markers:

<img width="836" height="102" alt="image" src="https://github.com/user-attachments/assets/16b074e9-30ec-4f98-b930-61dff0ac685b" />

klaxalk commented:
@MarlonRiv Hey, I have compiled everything and I am trying the simualation (https://github.com/ctu-mrs/mrs_multirotor_simulator/tree/ros2/tmux/mrs_one_drone). The drone takes off but refuses to fly anywhere (control through MRS status as well as simple "goto" is "outside of the safety area".

Commit:
formatting

MarlonRiv commented:
> @MarlonRiv hey, I am getting this in Rviz when visualizing the safety area coordinate markers:
> 
> <img alt="image" width="836" height="102" src="https://private-user-images.githubusercontent.com/1611176/517728030-16b074e9-30ec-4f98-b930-61dff0ac685b.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjM4MjU3MDcsIm5iZiI6MTc2MzgyNTQwNywicGF0aCI6Ii8xNjExMTc2LzUxNzcyODAzMC0xNmIwNzRlOS0zMGVjLTRmOTgtYjkzMC02MWRmZjBhYzY4NWIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MTEyMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTExMjJUMTUzMDA3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NGY0YjI0NDQ1ZmNjMmEwYzBiZTdkNTkxYmEwYTM3ZDExODUxMzBhZWZiMWQ4OTE3MjQ2ZDQ0NTZkNWU4MjRlNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.cBhoM5XuFnbRNd9yZxMjcpfEHgXB2hr71QopT7jkLrA">

I see, on my side i see it for just a little time but constantly, i will check it out

MarlonRiv commented:
> @MarlonRiv Hey, I have compiled everything and I am trying the simualation (https://github.com/ctu-mrs/mrs_multirotor_simulator/tree/ros2/tmux/mrs_one_drone). The drone takes off but refuses to fly anywhere (control through MRS status as well as simple "goto" is "outside of the safety area".

I will investigate, i am having errors as well that i didn't had before (I can't take off and control manager defaults to emergency controller and disables output). Would let you know once i found the cause! Thanks for letting me know

MarlonRiv committed:
Merge branch 'ros2' into ros2_safety_area_manager

MarlonRiv committed:
Fix of safety area validation services

MarlonRiv commented:
> > @MarlonRiv Hey, I have compiled everything and I am trying the simualation (https://github.com/ctu-mrs/mrs_multirotor_simulator/tree/ros2/tmux/mrs_one_drone). The drone takes off but refuses to fly anywhere (control through MRS status as well as simple "goto" is "outside of the safety area".
> 
> I will investigate, i am having errors as well that i didn't had before (I can't take off and control manager defaults to emergency controller and disables output). Would let you know once i found the cause! Thanks for letting me know

Hi @klaxalk ,  fixed! 

Fortunately, it was a silly mistake by wrong service remapping :// I thought it was a more complex problem by having a "recursive" service calling.  However, this bug led me to think if this service-based validation is the best approach. 

Right now, the ControlManager calls SafetyAreaManager services synchronously from within its own service callbacks to validate references, resulting in nested service calls: https://github.com/ctu-mrs/mrs_uav_managers/blob/ros2_safety_area_manager/src/control_manager/control_manager.cpp#L6630
It works as we have separate MutuallyExclusive callback groups + MultiThreadedExecutor.


I was thinking maybe we can avoid some service overhead/service complexity by directly validating in Control Manager. We could create an instance copy of the safety zone, `mrs_lib::safety_zone`, the shared library used in SafetyAreaManager, creating this copy from the information shared in the diagnostics topic from SafetyAreaManager: [Diagnostics msg](https://github.com/ctu-mrs/mrs_uav_managers/pull/25#:~:text=position_valid_3d%3A%20false-,border%3A,-horizontal_frame%3A%20world_origin)
-- And there would be no need to update this copy periodically but only when there is a change to be efficient.

Thought would be worth discussing it, maybe the service-based is enough, for the moment.

MarlonRiv commented:
> > @MarlonRiv hey, I am getting this in Rviz when visualizing the safety area coordinate markers:
> > <img alt="image" width="836" height="102" src="https://private-user-images.githubusercontent.com/1611176/517728030-16b074e9-30ec-4f98-b930-61dff0ac685b.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjM4MjU3MDcsIm5iZiI6MTc2MzgyNTQwNywicGF0aCI6Ii8xNjExMTc2LzUxNzcyODAzMC0xNmIwNzRlOS0zMGVjLTRmOTgtYjkzMC02MWRmZjBhYzY4NWIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MTEyMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTExMjJUMTUzMDA3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NGY0YjI0NDQ1ZmNjMmEwYzBiZTdkNTkxYmEwYTM3ZDExODUxMzBhZWZiMWQ4OTE3MjQ2ZDQ0NTZkNWU4MjRlNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.cBhoM5XuFnbRNd9yZxMjcpfEHgXB2hr71QopT7jkLrA">
> 
> I see, on my side i see it for just a little time but constantly, i will check it out

Hi @klaxalk  fixed! :)

Commit:
updated missing controller params, removed defaults from code

Commit:
updated the world file loading, new format

Commit:
added missing setting of default odom frame

MarlonRiv committed:
Simplified and fixes

Removed the global safety zone backup, not needed and just complicating
things.
Changes of world origin the safety area remains unchanged already

MarlonRiv committed:
Improving safety area manager

Cleaning and applying offset to safety area when the world origin
changes through estimation manager service-- only if it was previously
set with respect to world_origin (old world origin)

MarlonRiv committed:
Update new format of world config for tests

MarlonRiv committed:
Minor fix and changes

Commit:
wrong axis for up/down check

Commit:
fixed test for rc controled bumper

MarlonRiv committed:
Fix to avoid attempting transformations when safety area is disabled, and optimized validation callbacks

MarlonRiv committed:
Fix of using parameter from world file to enable/disable the safety area

Commit:
fixed and rerfactored tests against the new safety area manager

Commit:
SafetyArea: poliching, added loading of more configs
